src/resolve-names.cc

```diff
@@ -281,6 +298,67 @@ Result NameResolver::OnTeeLocalExpr(Expr* expr) {
   return Result::Ok;
 }
 
+Result NameResolver::BeginTryExpr(Expr* expr) {
+  try_blocks.push_back(expr);
+  return Result::Ok;
+}
+
+Result NameResolver::EndTryExpr(Expr*) {
+  try_blocks.pop_back();
+  return Result::Ok;
+}
+
+Result NameResolver::BeginTryBlockExpr(Expr* expr) {
+  PushLabel(&expr->try_block.block->label);
+  return Result::Ok;
+}
+
+Result NameResolver::EndTryBlockExpr(Expr* expr) {
+  PopLabel();
+  return Result::Ok;
+}
+
+Result NameResolver::BeginCatchExpr(Expr* expr) {
+  if (try_blocks.empty()) {
+    PrintError(&expr->loc, "Catch not in try block");
```

Comment:
The problem was I wasn't assuming that you could count of valid tree structure (just node construction). I realize now that I should have assumed the reader/parser are the only ones that need to check this.